pkg/redis/redisstore_test.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/storacha-network/indexing-service/pkg/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRedisStore(t *testing.T) {
+	ctx := context.Background()
+	testCases := []struct {
+		name       string
+		opts       []MockOption
+		behavior   func(t *testing.T, store *redis.Store[string, string])
+		finalState map[string]*redisValue
+	}{
+		{
+			name: "normal behavior",
+			behavior: func(t *testing.T, store *redis.Store[string, string]) {
+				store.Set(ctx, "key1", "value1", true)
+				store.Set(ctx, "key2", "value2", false)
+				store.Set(ctx, "key3", "value3", true)
+				store.Set(ctx, "key4", "value4", false)
+				store.SetExpirable(ctx, "key3", false)
+				store.SetExpirable(ctx, "key4", true)
+				require.Equal(t, "value1", testutil.Must(store.Get(ctx, "key1"))(t))
+				require.Equal(t, "value2", testutil.Must(store.Get(ctx, "key2"))(t))
+				require.Equal(t, "value3", testutil.Must(store.Get(ctx, "key3"))(t))
+				require.Equal(t, "value4", testutil.Must(store.Get(ctx, "key4"))(t))
+				_, err := store.Get(ctx, "key5")
+				require.ErrorIs(t, err, types.ErrKeyNotFound)
+			},
+			finalState: map[string]*redisValue{
+				"key1": {"value1", redis.DefaultExpire},
+				"key2": {"value2", 0},
+				"key3": {"value3", 0},
+				"key4": {"value4", redis.DefaultExpire},
+			},
+		},
+		{
+			name: "get errors",
+			opts: []MockOption{WithErrorOnGet(errors.New("something went wrong"))},
+			behavior: func(t *testing.T, store *redis.Store[string, string]) {
+				_, err := store.Get(ctx, "key1")
+				require.EqualError(t, err, "error accessing redis: something went wrong")
+			},
+		},
+		{
+			name: "set errors",
+			opts: []MockOption{WithErrorOnSet(errors.New("something went wrong"))},
+			behavior: func(t *testing.T, store *redis.Store[string, string]) {
+				err := store.Set(ctx, "key1", "value1", true)
+				require.EqualError(t, err, "error accessing redis: something went wrong")
+			},
+		},
+		{
+			name: "set expiration errors",
+			opts: []MockOption{WithErrorOnSetExpiration(errors.New("something went wrong"))},
+			behavior: func(t *testing.T, store *redis.Store[string, string]) {
+				err := store.SetExpirable(ctx, "key1", true)
+				require.EqualError(t, err, "error accessing redis: something went wrong")
+			},
+		},
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			mockRedis := NewMockRedis(testCase.opts...)
+			redisStore := redis.NewStore[string, string](
+				func(s string) (string, error) { return s, nil },
+				func(s string) (string, error) { return s, nil },
+				func(s string) string { return s },
+				mockRedis)
+			testCase.behavior(t, redisStore)
+			expectedFinalState := testCase.finalState
+			if expectedFinalState == nil {
+				expectedFinalState = make(map[string]*redisValue)
+			}
+			require.Equal(t, expectedFinalState, mockRedis.data)
+		})
+	}
+}
+
+type redisValue struct {
+	data    string
+	expires time.Duration
+}
+
+type MockRedis struct {
```

Comment:
There's a mock redis implementation here https://github.com/go-redis/redismock - is this not suitable?

Reply:
yea, I saw this -- generally, I want something a bit more stateful, that's simple and which I understand. I often that particular style of mocking library (seems analogous to this https://github.com/uber-go/mock) quite frustrating and awkward to use, and ultimately it doesn't tell you much. By and large, I just write my own mocks in go, cause I want to understand exactly the behavior that's happening under the hood and to see what comes out of that.

Reply:
there's also this -- https://github.com/alicebob/miniredis

ultimately I looked at both and was like "it would be easiest for me to write tests against something I expected to behave a very specific way.